File: bunny/bin/patron.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
A simple message consumer.

Usage: patron.py

It receives messages sent by a greeter of a worker.
It is the python version of java org.svom.bunny.worker.Patron.

"""

# First run 'pip install pika' if needed.
# See: https://www.rabbitmq.com/tutorials/tutorial-three-python.html

import sys
import pika
import getopt
import time
from datetime import date

def usage():
    """
    Prints the accepted options
    """
    sys.stdout.write(__doc__)
    sys.exit(1)

def process():
    """
    Initializes the RabbitMQ stuff, wait for messages, prints them.

    """
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host='hurukan'))
    channel = connection.channel()
    channel.exchange_declare(exchange='bunny', type='fanout',
                             durable=False, auto_delete=True)
    
    result = channel.queue_declare(durable=False, exclusive=False,
                                   auto_delete=True)
    queue_name = result.method.queue

    channel.queue_bind(exchange='bunny', queue=queue_name)

    def callback(canal, method, properties, body):
        
        print("Message: %s" % body)
        try:
            i = body.index('=')
            content = body[(i+1):].strip()
        except:
            content = body
            
        print("Content: %s" % content)
        if content == 'close':
            channel.close()
            connection.close()
            print 'Patron is dead !'

    print 'Patron is listening ...'
    channel.basic_consume(callback, queue=queue_name, no_ack=True)
    channel.start_consuming()
    
def main(argv):

    try:                                
        opts, args = getopt.getopt(argv, 'h', ['help'])

    except getopt.GetoptError:          
          usage()                         
   
    for opt, val in opts:
        if opt in ('-h', '--help'):
            usage()
       
    process()

if __name__ == '__main__':
    main(sys.argv[1:])

#___________________________________________________________________________

File: bunny/src/main/java/org/svom/bunny/worker/Patron.java
//______________________________________________________________________________

//      Svom  workbench - Bunny project
//______________________________________________________________________________

package org.svom.bunny.worker;

import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Consumer;
import com.rabbitmq.client.DefaultConsumer;
import com.rabbitmq.client.Envelope;
import com.rabbitmq.client.AMQP;

import java.io.IOException;
import java.util.concurrent.TimeoutException;

/**
 * A very simple message consumer.
 *
 * It receives messages sent by a greeter or a worker.
 *
 * Usage: java org.svom.bunny.worker.Patron
 *
 * @see Greeter
 * @see https://www.rabbitmq.com/tutorials/tutorial-one-java.html
 * @author Jean-Paul Le Fèvre <jean-paul.lefevre@cea.fr>
 */
public final class Patron {
    /**
     * The name of the exchange.
     */
    private static final String EXCHANGE_NAME = "bunny";
     /**
     * The name of the message queue.
     */
    private static final String QUEUE_NAME = "patron_queue";

    //__________________________________________________________________________
    /**
     * Creates the object.
     */
    private Patron() {

        System.out.println("Patron created !");
    }
    //__________________________________________________________________________
    /**
     * Main method.
     *
     * @param args the arguments of the program.
     * @throws IOException if something goes wrong.
     * @throws TimeoutException if something goes wrong.
     */
    public static void main(String[] args)
        throws IOException, TimeoutException {

        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost("hurukan");
        Connection connection = factory.newConnection();
        Channel channel = connection.createChannel();
        channel.exchangeDeclare(EXCHANGE_NAME, "fanout", false, true, null);

        // Not durable, not exclusive, autoDelete
        String str = channel.queueDeclare(QUEUE_NAME,
                                          false, false, true, null).getQueue();
        channel.queueBind(str, EXCHANGE_NAME, "");

        System.out.println("Waiting for messages on " + str + " ...");

        Consumer consumer = new DefaultConsumer(channel) {
                @Override
                public void handleDelivery(String consumerTag,
                                           Envelope envelope,
                                           AMQP.BasicProperties properties,
                                           byte[] body)
                    throws IOException {

                    String message = new String(body, "UTF-8");
                    System.out.println("Got '" + message + "'");
                }
            };

        channel.basicConsume(QUEUE_NAME, true, consumer);
    }
    //__________________________________________________________________________
}

